File: Data Structures & Algorithms/find-target-in-rotated-sorted-array/submission-2.cpp
/**
 * 1. length-n sorted array (asending order)
 * 2. rotated 1~n times: n times -> unchanged
 * 3. all elements are unique
 * 4. O(log n) solution
 * Given: rotated sorted array ``nums`` and an ``target``.
 * Find: index of ``target`` within ``nums``.
 */

 /**
  * O(log n) Solution: bineary search variation
  * `[3, 4, 5, 1, 2]` & `4` -> first half
  * `[3, 4, 5, 1, 2]` $ `1` -> second half
  * `[4, 5, 1, 2, 3]` $
  */

class Solution {
public:
    int search(vector<int>& nums, int target) {
        int left = 0;
        int right = nums.size() - 1;
        while (left <= right) {
            // Prevent overflow
            int mid = left + (right - left) / 2;
            if (nums[mid] == target) {
                return mid;
            } 
            
            // Left half [left, mid] is sorted
            if (nums[left] <= nums[mid]) {
                // Is target inside the sorted left half?
                if (nums[left] <= target && target < nums[mid]) {
                    right = mid - 1;
                } else {
                    left = mid + 1;
                }
            }

            // Otherwise right half [mid, right] is sorted
            else {
                // Is target inside the sorted right half?
                if (nums[mid] < target && target <= nums[right]) {
                    left = mid + 1;
                } else {
                    right = mid - 1;
                }
            }

        }
        return -1;
    }
};

File: Data Structures & Algorithms/find-minimum-in-rotated-sorted-array/submission-2.py
class Solution:
    def findMin(self, nums: List[int]) -> int:
        """
        O(logn) solution for sorted array -> binary search

        Where can minimum element can reside?
        1. No Rotation:
            nums[0] < nums[-1] -> min = 0
        2. Rotated: nums[min] < nums[min - 1]
            2.1 [l, m]
            2.2 [m, r]
        """
        res = nums[0]
        l, r = 0, len(nums) - 1

        while l <= r:
            if nums[l] < nums[r]:
                res = res = min(res, nums[l])
            m = l + (r - l) // 2
            res = min(res, nums[m])

            # sorted left half -> right half
            if nums[l] <= nums[m]:
                l = m + 1

            # sorted right half -> left half
            else:
                r = m - 1

        return res


